### DISCUSSION TITLE
Does the community need this?

### DISCUSSION THREAD

**Discussion by ramsey** [php-identifier/identifier#10] [General]
We got into a discussion on Twitter[^1] about whether there's a need for these interfaces. Maybe there's not.

My primary goal in creating this is to transition folks away from using interfaces in the `Ramsey` namespace (i.e., `Ramsey\Uuid\UuidInterface`) to common, neutral interfaces. I want to decouple the ecosystem's reliance on a single UUID package and allow others to create identifier types that can be used in the same places, without being tied to a specific format.

There are UUIDs (8 standard types), ULIDs, Snowflake IDs, MongoDB ObjectIds, Flickr Tickets, and, with the new uuidrev working group at the IETF, there will likely be a few more UUID types. All of these types of IDs solve similar problems, but choosing one is a matter of application (architecture, storage, etc.), and sometimes organization (cost, etc.), need. A common interface for generating and passing around IDs would abstract away the creation and introspection of identifiers, so a development team can make decisions about the types of IDs to use, without locking themselves into any particular type. There's still some lock-in at the storage level, but the IDs can be mostly opaque to the code that uses the interfaces. This also means that the same code can be reused to work with different types of IDs, such as might be the case if using a library to work with different types of data from different data stores.

Additionally, the interfaces make me feel like less of a linchpin in the ecosystem, allowing others to release their own ID libraries that teams can use as drop-in replacements, should I ever find myself in a position where I'm unable to continue work on ramsey/uuid. Though, as @mathiasverraes pointed out[^2], maybe this isn't a compelling argument, since anyone can fork a library and search-and-replace the namespaces to solve this problem.

Thoughts? Is this something the community needs?

[^1]: https://twitter.com/mathiasverraes/status/1576670888053067776
[^2]: https://twitter.com/mathiasverraes/status/1576917807446765575

**Comment by ralflang** [comment]
I wouldn't worry too much about the exact vendor namespace. Also, provided you don't actively retract existing versions, relying on your implementation with sane version constraints should not be much of an issue.

I cannot tell if "the community needs" this. So far, my identifiers are generated strings and they collapse to string early in their lifecycle - mostly before the identified object even exists. They don't really mutate over lifetime. That might not be how "the community" as a whole does it and it may make sense to have an identifier as a typed field.

But why so defensive? This is a fine UUID handling library. If anything at all, the accompanying interface demands a little much. As described above, Stringable would be what an identifier type must be, Serializable is overkill for me. While it might be tricky to deduce the type (and semantics) of an ID from its string and easy from its more elaborate JSON serialization, I also rarely have that need.

**Reply by ramsey** [reply]
> But why so defensive?

I don’t think I’m being defensive about anything. Where do I come across as defensive?

**Reply by ralflang** [reply]
Maybe just a perception thing on my side. Not my native language. 290 million installs of /uuid show quite some community interest and that will continue for /identifier.

**Reply by ramsey** [reply]
This discussion is for the `identifier/identifier` interfaces. I’m not questioning whether the community needs the `ramsey/identifier` package, which is separate but depends on these interfaces. I apologize for the confusion.

**Comment by MGatner** [comment]
I think this is a great addition! 

Something you didn't mention: UUIDs and ULID (and probably most of those other formats) generation is inherently an impure process, relying on random numbers and/or system time. This means in order to generate new IDs in your core code you should pass a factory by (surprise!) an abstraction.

Implied in this process to me is the need for identifier factory interfaces in every project, which makes it a great candidate for a library; allowing interoperability among various providers is added bonus.
